### DISCUSSION TITLE
io uring CPU 100%

### DISCUSSION THREAD

**Discussion by calvin2021y** [libuv/libuv#4032] [General]
after upgrade to 1.45, the linux cpu  often run at 100%.  (cause other app slow)

Is this normal ?  any way to reduce the case ?

**Comment by calvin2021y** [comment]
`strace` show most of call is :

```sh
io_uring_enter(5, 10, 10, IORING_ENTER_GETEVENTS, NULL, 0) = 10
epoll_pwait(3, [{events=EPOLLHUP, data={u32=24, u64=24}}, {events=EPOLLHUP, data={u32=16, u64=16}}, {events=EPOLLHUP, data={u32=17, u64=17}}, {events=EPOLLHUP, data={u32=21, u64=21}}, {events=EPOLLHUP, data={u32=18, u64=18}}, {events=EPOLLHUP, data={u32=20, u64=20}}, {events=EPOLLHUP, data={u32=21, u64=21}}, {events=EPOLLHUP, data={u32=22, u64=22}}, {events=EPOLLHUP, data={u32=23, u64=23}}, {events=EPOLLHUP, data={u32=25, u64=25}}], 1024, -1, NULL, 8) = 10
io_uring_enter(5, 10, 10, IORING_ENTER_GETEVENTS, NULL, 0) = 10
epoll_pwait(3, [{events=EPOLLHUP, data={u32=24, u64=24}}, {events=EPOLLHUP, data={u32=16, u64=16}}, {events=EPOLLHUP, data={u32=17, u64=17}}, {events=EPOLLHUP, data={u32=21, u64=21}}, {events=EPOLLHUP, data={u32=18, u64=18}}, {events=EPOLLHUP, data={u32=20, u64=20}}, {events=EPOLLHUP, data={u32=21, u64=21}}, {events=EPOLLHUP, data={u32=22, u64=22}}, {events=EPOLLHUP, data={u32=23, u64=23}}, {events=EPOLLHUP, data={u32=25, u64=25}}], 1024, -1, NULL, 8) = 10
io_uring_enter(5, 10, 10, IORING_ENTER_GETEVENTS, NULL, 0) = 10
epoll_pwait(3, [{events=EPOLLHUP, data={u32=24, u64=24}}, {events=EPOLLHUP, data={u32=16, u64=16}}, {events=EPOLLHUP, data={u32=17, u64=17}}, {events=EPOLLHUP, data={u32=21, u64=21}}, {events=EPOLLHUP, data={u32=18, u64=18}}, {events=EPOLLHUP, data={u32=20, u64=20}}, {events=EPOLLHUP, data={u32=21, u64=21}}, {events=EPOLLHUP, data={u32=22, u64=22}}, {events=EPOLLHUP, data={u32=23, u64=23}}, {events=EPOLLHUP, data={u32=25, u64=25}}], 1024, -1, NULL, 8) = 10
```

**Reply by bnoordhuis** [reply]
Can you post the output of `uname -a` and `strace -yy <program>`?

**Reply by calvin2021y** [reply]
uname 

```sh
Linux 6.1.27 #8 
```
the strace is huge, and not goto 100% after start(take some time to trigger this)

now I disable uring by UV_USE_IO_URING, try watch if problem still exists

**Comment by calvin2021y** [comment]
after disable io uring with `UV_USE_IO_URING`,  5 hours CPU time is `0:00.35`. 

with io uring,  CPU time is almost like 100%.   

the app use libuv for curl fd pull.

**Comment by santigimeno** [comment]
Can you check again disabling the `ctl` ring (comment this line: https://github.com/libuv/libuv/blob/v1.x/src/unix/linux.c#L565) ?

**Reply by calvin2021y** [reply]
confirm no CPU 100% when remove ` uv__iou_init(loop->backend_fd, &lfields->ctl, 256, 0);`.

**Reply by santigimeno** [reply]
Adding the `ctl` ring back, can you post the `strace -yy -p pid`  output somewhere (only when the process goes 100%)? Also, if possible,  in a debug build, can you attach the debugger to the process when goes 100% and post multiple backtraces (just in case it's hung in a loop somewhere) Thanks!

**Reply by bnoordhuis** [reply]
@calvin2021y did you ever get to the bottom of this? Maybe it's a libuv bug, maybe it's a kernel bug that libuv can work around.

**Reply by calvin2021y** [reply]
sorry for late reply.

The `strace` result size is huge,  I get lost in the logs. I will try more and update.
